### PR TITLE
SECZ-1582: Add Linux support

### PR DESCRIPTION
**Pull request by rbamos**
Testing:
- [x] Tested locally at the repo root
- [x] Tested locally one step up from repo root (to check that utils.sh fetch works)
- [x] Tested install on Linux box
- [x] Tested install via Ansible; see https://github.com/panorama-ed/ansible-services/pull/5
- [x] Started configuration, confirmed that Leapp CLI correctly communications with Leapp
- [x] Fully configured a Leapp installation with config.sh

Core objective: make this script work on both Linux & Mac

Additional tweaks:
* Utils script is correctly fetched when not present
* Config broken out into a separate step to facilitate Ansible installation
* Non-interactive mode created to facilitate Ansible installation
* Small cleanup

### CODE REVIEW

File: setup.sh [hunk 1]
```diff
@@ -1,133 +1,135 @@
...
+            group=$(id -gn | cut -d ' ' -f 1)
+        fi
+        sudo chown -R "$(whoami):$group" "/home/$(whoami)"
+    fi
+
+    NONINTERACTIVE=1 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
 
     # If using an M1 machine, load shell environment to run brew commands
-    if [[ $(uname -m) == 'arm64' ]]; then
-        echo ‘# Set PATH, MANPATH, etc., for Homebrew.’ >> ~/.zprofile
```
Comment (rbamos) [inline setup.sh:32]: Unicode Quote -> ASCII quote

File: setup.sh [hunk 2]
```diff
@@ -1,133 +1,135 @@
...
-        echo ‘eval "$(/opt/homebrew/bin/brew shellenv)"’ >> ~/.zprofile
+    if [[ "$kernel_name" == "Darwin" ]] && [[ $(uname -m) == 'arm64' ]]; then
+        echo '# Set PATH, MANPATH, etc., for Homebrew.' >> ~/.zprofile
+        echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> ~/.zprofile
         eval "$(/opt/homebrew/bin/brew shellenv)"
     fi
+
+    if [[ "$kernel_name" == "Linux" ]]; then
+        (echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> ~/.bashrc
```
Comment (eswidler) [inline setup.sh]: Do we need to modify ~/.bashrc?  I know it looks like we did something similar with the mac version of the script.  But if we're only loading shell env vars to be able to run "brew" commands for the duration of this script, is it really necessary to modify our default shell behavior?

Could we just drop this line and use `eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"` below?

Comment (eswidler) [reply setup.sh]: Also Linuxbrew is not installed currently on the eng image.  How can we ensure that linuxbrew is installed?

**Comment by eswidler** [comment]
An overall comment:
One thing I'm learning is that when logged in as `administrator`, the home directory does not exist.  So anything using `~/...` would reference a non-existing dir.

How are you imagining this is run?  As an `AnsibleAdmin` user that still has it's own `/home` dir & sudo-level access?

File: setup.sh [hunk 3]
```diff
@@ -1,133 +1,135 @@
...
+        eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
+        sudo apt install -y build-essential
+        brew install gcc
```
Comment (eswidler) [inline setup.sh]: Could we do `sudo apt install gcc` instead? If we can install gcc without brew, we may not need `linuxbrew` at all.

**Comment by eswidler** [comment]
Any `apt` package will be installed on the root volume and not needed if we're creating another workspace based on an image using that volume.  Given that, do we need to have all of the `sudo apt install PACKAGE` lines we have?  Could we instead rely on these packages already being available on the root volume? 

EDIT: I remember that we have the uninstall process here for testing the setup from scratch. That would require reinstalling the `apt` packages when reinstalling.  I'm ok with leaving these installation calls for now even though they're likely redundant on ansible.  Seems low priority to resolve for now.

File: setup.sh [hunk 4]
```diff
@@ -1,133 +1,135 @@
...
+        curl https://asset.noovolari.com/latest/Leapp_0.26.1_amd64.deb -o ~/Downloads/leapp.deb
+        sudo dpkg -i ~/Downloads/leapp.deb
+        sudo mv /usr/bin/leapp /usr/bin/leapp-desktop
     fi
+    curl -fsSL https://deb.nodesource.com/setup_22.x -o ~/Downloads/nodesource_setup.sh
+    sudo bash ~/Downloads/nodesource_setup.sh
+    sudo apt install -y nodejs
+    sudo apt install -y npm
+    sudo npm install -g @noovolari/leapp-cli
+    # brew install Noovolari/brew/leapp-cli
```
Comment (eswidler) [inline setup.sh]: Can we drop this?

File: rollback_setup.sh [hunk 5]
```diff
@@ -7,17 +7,36 @@ while true; do
...
+    sudo npm uninstall -g @noovolari/leapp-cli
+    # Remove node
+    sudo apt remove -y npm
+    sudo apt remove -y nodejs
+    # Remove leapp
+    sudo dpkg -r leapp
+    sudo dpkg -P leapp
+    # Remove session-manager-plugin
+    sudo dpkg -r session-manager-plugin
+    sudo dpkg -P session-manager-plugin
```
Comment (eswidler) [inline rollback_setup.sh:38]: Can you combine the remove and purge flags to run this once?

Comment (eswidler) [inline rollback_setup.sh:38]: Can you combine the remove and purge flags to run this once for each package?

Comment (rbamos) [reply rollback_setup.sh:38]: No, dpkg complains if you do

File: setup.sh (hunk 2 above)
Comment (rbamos) [reply setup.sh]: My understanding is `eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"` sets up the current shell session to run brew, so you need to add it to `.bashrc` for it to apply to every shell session

File: setup.sh (hunk 3 above)
Comment (rbamos) [reply setup.sh]: We could. Actually this code is unreachable and we could just remove it outright too -- the Brew installation if check above first checks if we're on Mac OS, so if we're on Linux Brew doesn't get installed and this code doesn't run

Comment (rbamos) [reply setup.sh]: Do we need Linuxbrew for other purposes?

File: setup.sh (hunk 4 above)
Comment (rbamos) [reply setup.sh]: Yup, not needed

**Comment by rbamos** [comment]
> An overall comment:
> One thing I'm learning is that when logged in as administrator, the home directory does not exist. So anything using ~/... would reference a non-existing dir.

> How are you imagining this is run? As an AnsibleAdmin user that still has it's own /home dir & sudo-level access?

This runs as the `ansible` user. I did add

```
  if [[ "$kernel_name" == "Linux" ]] && [[ ! -e "/home/$(whoami)" ]]; then
    sudo mkdir -p "/home/$(whoami)"
    if id -gn | grep 'users' > /dev/null; then
        group='users'
    else
        group=$(id -gn | cut -d ' ' -f 1)
    fi
    sudo chown -R "$(whoami):$group" "/home/$(whoami)"
  fi
```

But this doesn't actually get run. I could add move it to a spot where it would get run

File: setup.sh (hunk 3 above)
Comment (eswidler) [reply setup.sh]: I haven't run anything setting up the eng image that required Linuxbrew.

**Review by eswidler — APPROVED** [review]

**Review by eswidler — APPROVED** [review]

**Pull request merged**